tests/pytest/test_binary_value.py

```diff
@@ -0,0 +1,244 @@
+        pytest.fail("Expected ValueError when assigning negative number to unsigned BinaryValue")
+
+
+def test_init_not_enough_bits(capsys):
+    bin1 = BinaryValue(value=128, n_bits=7, bigEndian=True, binaryRepresentation=BinaryRepresentation.UNSIGNED)
+    assert bin1._str == "0000000"
+    assert bin1.binstr == "0000000"
+    assert bin1.integer == 0
+    captured = capsys.readouterr()
+    assert captured.out == "WARNING: truncating value to match requested number of bits (8 -> 7)\n"
+
+    bin2 = BinaryValue(value="1111110000101100", n_bits=12, bigEndian=False,
+                       binaryRepresentation=BinaryRepresentation.TWOS_COMPLEMENT)
+    assert bin2._str == "110000101100"
+    assert bin2.binstr == "110000101100"
+    assert bin2.integer == -980
+
+    bin3 = BinaryValue(value="1111110000101100", n_bits=11, bigEndian=False,
+                       binaryRepresentation=BinaryRepresentation.SIGNED_MAGNITUDE)
+    assert bin3._str == "10000101100"
+    assert bin3.binstr == "10000101100"
+    assert bin3.integer == -44
+
+
+def test_init_short_binstr_value():
+    bin1 = BinaryValue(value="0", n_bits=4, bigEndian=True, binaryRepresentation=BinaryRepresentation.UNSIGNED)
+    assert bin1._str == "0000"
+    assert bin1.binstr == "0000"
+    assert bin1.integer == 0
+
+    bin2 = BinaryValue(value="Z", n_bits=8, bigEndian=False, binaryRepresentation=BinaryRepresentation.UNSIGNED)
+    assert bin2._str == "0000000Z"
+    assert bin2.binstr == "0000000Z"
+    with pytest.raises(ValueError):
+        assert bin2.integer == 0
+        pytest.fail("Expected ValueError when resolving Z to integer")
+
+    bin3 = BinaryValue(value="01", n_bits=8,
+                       bigEndian=False, binaryRepresentation=BinaryRepresentation.SIGNED_MAGNITUDE)
+    assert bin3._str == "00000001"
+    assert bin3.binstr == "00000001"
+    assert bin3.integer == 1
+
+    bin4 = BinaryValue(value="1", n_bits=8,
+                        bigEndian=True, binaryRepresentation=BinaryRepresentation.SIGNED_MAGNITUDE)
+    # 1 digit is too small for Signed Magnitude representation, so setting binstr will fail, falling back to buff
+    bin4._str == "10000000"
+    bin4.binstr == "10000000"
+    bin4.integer == 1
+
+def test_defaults():
+    bin1 = BinaryValue(17)
+    assert bin1.binaryRepresentation == BinaryRepresentation.UNSIGNED
+    assert bin1.big_endian == True
+    assert bin1._n_bits == None
+    assert bin1.integer == 17
+
+def test_indexes():
+    bin1 = BinaryValue(value=-980, n_bits=32,
+                       bigEndian=False, binaryRepresentation=BinaryRepresentation.TWOS_COMPLEMENT)
+
+    bin2 = bin1[3:2]
+    assert bin2.binstr == "11"
+    assert bin2.integer == -1
+
+    with pytest.raises(IndexError):
+        bin3 = bin1[32]
+
+    with pytest.raises(IndexError):
+        bin4 = bin1[1:2]
+
+    with pytest.raises(IndexError):
+        bin5 = bin1[-1:4]
+
+    with pytest.raises(IndexError):
+        bin6 = bin1[2:-2]
```

Comment:
If it raises an exception there's no point naming the result, is there?
```suggestion
    with pytest.raises(IndexError):
        bin1[32]

    with pytest.raises(IndexError):
        bin1[1:2]

    with pytest.raises(IndexError):
        bin1[-1:4]

    with pytest.raises(IndexError):
        bin1[2:-2]
```